libcxx/include/__tree

```diff
@@ -717,6 +718,59 @@ private:
   friend class __tree_const_iterator;
 };
 
+template <class _Reference, class _EndNodePtr, class _NodePtr, class _Func, class _Proj>
+_LIBCPP_HIDE_FROM_ABI bool __tree_iterate_from_root(_EndNodePtr __last, _NodePtr __root, _Func& __func, _Proj& __proj) {
```

Comment:
This could take a predicate called something like `_EarlyExit` or `_Break` or something like that. And then `__tree_iterate_from_begin` can call that with `[](auto node) { return node == last; }`.

And the `ranges::for_each` implementation can just use `__tree_iterate_from_root([](auto) { return false; })`.